res/controllers/Numark-Mixtrack-Platinum-scripts.js

```diff
@@ -0,0 +1,1418 @@
+            }
+        }
+    };
+    this.key_up = new components.Button({
+        inKey: 'pitch_up',
+        direction: 1,
+        input: key_up_or_down,
+    });
+    this.key_down = new components.Button({
+        inKey: 'pitch_down',
+        direction: -1,
+        input: key_up_or_down,
+    });
+    this.key_up.other = this.key_down;
+    this.key_down.other = this.key_up;
+
+    loop_base = function(midino, obj) {
+        return _.assign({
+            midi: [0x94 + midi_chan, midino],
+            on: 0x01,
+            sendShifted: true,
+            shiftChannel: true,
+            shiftOffset: -0x10,
+        }, obj);
+    };
+
+    this.alternate_manloop = new components.ComponentContainer({
+        loop_in: new components.HotcueButton(loop_base(0x38, {
+            number: 5,
+        })),
+        loop_out: new components.HotcueButton(loop_base(0x39, {
+            number: 6,
+        })),
+        loop_toggle: new components.HotcueButton(loop_base(0x32, {
+            number: 7,
+        })),
+        // there are two hotcue 8 controls, one for the loop_halve button and
+        // one for the loop_double button. These buttons are two individual
+        // functions when in manual loop mode, but they behave as one button
+        // in hotcue mode.
+        loop_halve: new components.HotcueButton(loop_base(0x34, {
+            number: 8,
+        })),
+        loop_double: new components.HotcueButton(loop_base(0x35, {
+            number: 8,
+        })),
+    });
+    this.normal_manloop = new components.ComponentContainer({
+        loop_in: new components.Button(loop_base(0x38, {
+            inKey: 'loop_in',
+            outKey: 'loop_start_position',
+            outValueScale: function (value) {
+                return (value != -1) ? this.on : this.off;
+            },
+        })),
+        loop_out: new components.Button(loop_base(0x39, {
+            inKey: 'loop_out',
+            outKey: 'loop_end_position',
+            outValueScale: function (value) {
+                return (value != -1) ? this.on : this.off;
+            },
+        })),
+        loop_toggle: new components.LoopToggleButton(loop_base(0x32, {})),
+        loop_halve: new components.Button(loop_base(0x34, {
+            key: 'loop_halve',
+            input: function(channel, control, value, status) {
+                if (this.isPress(channel, control, value, status)) {
+                    engine.setValue(deck.currentDeck, "loop_scale", 0.5);
+                }
+            },
+        })),
+        loop_double: new components.Button(loop_base(0x35, {
+            key: 'loop_double',
+            input: function(channel, control, value, status) {
+                if (this.isPress(channel, control, value, status)) {
+                    engine.setValue(deck.currentDeck, "loop_scale", 2.0);
+                }
+            },
+        })),
+    });
+    // swap normal and alternate manual loop controls
+    if (UseManualLoopAsCue) {
```

Comment:
I figure most users won't even know about the extra options and in that case, the controller will work very similar to how the manual it comes with says it should. I tried to make all the defaults unsurprising. But when I play, I rarely use the autoloop functionality, so for me it makes sense to make that control cue points by default.

For the sampler, it probably makes sense to control 8 banks by default, but this might be unexpected to a user, so I want them to opt into it explicitly.

Reply:
Going to leave as is. This will be less of an issue when we have a UI for controller options. The code will be just as complex, but the burden on the user will be less. I still think having all the options available is useful for users who want to customize their workflow.